### PR TITLE
Add 2 methods: fetch all threads, fetch all users involved in threads

### PR DESCRIPTION
**Pull request by 2FWAH**
As fetchAllUsers() seems broken (#282 ), I added two methods:

- fetchThreads(thread_location) to fetch all threads in ThreadLocation
- fetchAllUsersFromThreads(threads) to fetch all users involved in threads.

Example: to fetch all users in INBOX simply use:
Users = client.fetchAllUsersFromThreads(client.fetchThreads(ThreadLocation.INBOX))

fetchAllUsersFromThreads() might be long to finish, should progress be printed? (based on the number of threads given as parameter)

### CODE REVIEW

File: fbchat/client.py [hunk 1]
```diff
@@ -472,7 +472,51 @@ def resetDefaultThread(self):
...
+    def fetchThreads(self, thread_location):
+        """
+        Get all threads in thread_location.
 
+        :param thread_location: models.ThreadLocation: INBOX, PENDING, ARCHIVED or OTHER
+        :return: :class:`models.Thread` objects
+        :rtype: list
+        :raises: FBchatException if request failed
+        """
+        Threads = []
```
[the remark this answers is not in the document]
Comment (2FWAH) [reply fbchat/client.py]: I fixed the code to use snake case convention.
What do you mean by no need to initialize it to [] ?

Like `threads = list()` ?

File: fbchat/client.py [hunk 2]
```diff
@@ -472,7 +472,51 @@ def resetDefaultThread(self):
...
+        Threads += self.fetchThreadList(thread_location=thread_location)
+        if len(Threads) == 0:
+            return []
+        while True:
+            lastThreadTimestamp = Threads[-1].last_message_timestamp
+            candidates = self.fetchThreadList(before=lastThreadTimestamp, thread_location=thread_location)  # return at max 20 threads before lastThreadTimestamp (included)
```
[the remark this answers is not in the document]
Comment (2FWAH) [reply fbchat/client.py]: last_message_timestamp is a millisecond timestamp, it's unlikely that 2 timestamps are the same.
But what happens if FB decides to round the timestamp to the second?
Doing with `if len(candidates) > 1:` adds one request, but looks more _clean_, no ? (-1 looks like a trick)

**Comment by 2FWAH** [comment]
> it would be nice to have a limit parameter

Yes, BTW an `after` parameter would be useful to fetch threads after a timestamp.
I think by default no _limit_ or _after_ setting should be set, right?

File: fbchat/client.py [hunk 3]
```diff
@@ -472,7 +472,51 @@ def resetDefaultThread(self):
...
+        """
+        Users = []
+        for thread in threads:
+            if thread.type == ThreadType.USER:
+                if thread.uid not in [user.uid for user in Users]:
+                    Users.append(thread)
+            elif thread.type == ThreadType.GROUP:
+                for userID in thread.participants:
+                    if userID not in [user.uid for user in Users]:
+                        Users.append(self.fetchUserInfo(userID)[userID])
```
[the remark this answers is not in the document]
Comment (2FWAH) [reply fbchat/client.py]: Done, but keep in mind that the order of users is no longer the order of discovery.

File: fbchat/client.py [hunk 4]
```diff
@@ -472,7 +472,54 @@ def resetDefaultThread(self):
     """
     FETCH METHODS
     """
+    
+    def fetchThreads(self, thread_location, after=None, limit=None):
```
Comment (madsmtm) [inline fbchat/client.py]: Yes, `limit` and `after` should be `None` by default

**Comment by 2FWAH** [comment]
Totally forgot about this PR! :-1: 
Still relevant today with the refactoring of the API (aka V2.0.0)?

**Comment by madsmtm** [comment]
I'd say it's still relevant, `v2.0.0` is still quite far away ;)

**Comment by madsmtm** [comment]
And besides, there's nothing preventing us from using part of the implementation from here in the new API

**Comment by 2FWAH** [comment]
Done, testing needed :)
Support for datetime objects for before and after parameters is not yet implemented.
Should be part of #278 no?

By the way is it possible to know last modification date of some threads present in the test accounts?
That would be useful to add tests for before and after parameters.

File: fbchat/client.py [hunk 5]
```diff
@@ -487,6 +487,82 @@ def _forcedFetch(self, thread_id, mid):
...
+            # FB returns a sorted list of threads
+            if (before is not None and int(last_thread_timestamp) > before) or \
+                (after is not None and int(last_thread_timestamp) < after):
+                break
+
+        # Return only threads between before and after (if set)
+        if before is not None or after is not None:
+            for t in list(threads):
+                last_message_timestamp = int(t.last_message_timestamp)
+                if (before is not None and last_message_timestamp > before) or \
```
Comment (madsmtm) [inline fbchat/client.py]: It seems kinda unclean that this is duplicated further up, is there a way we can refactor, and merge these?

**Comment by madsmtm** [comment]
You're right about not using datetime objects yet, at least that'll give some sort of consistency ;)

But no, I can't think of a way to properly test these methods

**Comment by madsmtm** [comment]
I went ahead and did some of the changes I requested, since this PR would have gotten lost otherwise (and I really wanted this functionality :grin:). Thanks for your work! 🎉

**Pull request merged**